From d5573652a0b8bc0492a5cc6bd465abbb15165ec4 Mon Sep 17 00:00:00 2001
From: Kevin Hao <kexin.hao@windriver.com>
Date: Tue, 5 Mar 2013 18:23:36 +0800
Subject: [PATCH 525/547] powerpc/85xx: release all the secondary cpus in kexec
 boot

In kexec boot, all the secondary cpus are spinning on a common
spinloop. So we should release them first before invoking
smp_generic_kick_cpu to kick off a specific cpu.

Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
Signed-off-by: Bin Jiang <bin.jiang@windriver.com>
---
 arch/powerpc/platforms/85xx/smp.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/arch/powerpc/platforms/85xx/smp.c b/arch/powerpc/platforms/85xx/smp.c
index 5e46838..f3a33e1 100644
--- a/arch/powerpc/platforms/85xx/smp.c
+++ b/arch/powerpc/platforms/85xx/smp.c
@@ -278,6 +278,7 @@ static int __cpuinit smp_85xx_kick_cpu(int nr)
 #endif
 #ifdef CONFIG_PPC64
 	unsigned long *ptr = NULL;
+	static int secondary_cpus_released;
 #endif
 
 	WARN_ON(nr < 0 || nr >= NR_CPUS);
@@ -408,10 +409,16 @@ static int __cpuinit smp_85xx_kick_cpu(int nr)
 #else
 	ptr  = (unsigned long *)((unsigned long)&__run_at_kexec);
 	/* We shouldn't access spin_table from the bootloader to up any
-	 * secondary cpu for kexec kernel, and kexec kernel already
-	 * know how to jump to generic_secondary_smp_init.
+	 * secondary cpu for kexec kernel, all secondary cpus are spinning
+	 * on a common spinloop. So we should release them by invoking
+	 * smp_release_cpus
 	*/
-	if (!*ptr) {
+	if (*ptr) {
+		if (!secondary_cpus_released) {
+			smp_release_cpus();
+			secondary_cpus_released = 1;
+		}
+	} else {
 		flush_spin_table(spin_table);
 		out_be32(&spin_table->pir, hw_cpu);
 		out_be32(&spin_table->addr_h,
-- 
1.8.4.93.g57e4c17

